proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,184 @@
+therefore requiring access to this sysvar data via syscalls and *not* via
+providing the account directly. Requiring the account would break the programs'
+ABIs.
+
+### Reducing Syscall Bloat
+
+Currently, sysvar data layouts are very difficult to change. When called, the
+syscall handler will copy the entire sysvar to the given memory location.
+
+Notably, BPF programs do not inform the syscall handler of the expected size
+of the sysvar. BPF programs thus effectively assume that the requested sysvars
+will forever have the same size.
+
+This prevents future upgrades that shrink or extend sysvars. In the same
+spirit, if a new sysvar is created, a new syscall must be created.
+
+Furthermore, if certain queries are requred by BPF programs - such as
+`SlotHashes::get_slot()` and `StakeHistory::get_entry()` - those single-element
+retrievals will *also* require new syscalls.
+
+It's worth noting that at the time of this writing, the work required for
+[SIMD-0088](./0088-enable-core-bpf-programs.md) will require at least five new
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+## Detailed Design
+
+By simply redefining the `get` API to accept `offset` and `length` parameters,
+one single `get` syscall can be used by all sysvars to retrieve data.
+
+```rust
+fn get(offset: usize, length: usize) -> &[u8];
+```
+
+The specifics of returning optional values or errors can be left up to the
+implementation, but the following cases would be handled with errors:
+
+- The sysvar data is unavailable.
+- The sysavr data is corrupt.
+- The provided offset or length would be out of bounds on the sysvar data.
```

Comment:
it occurs to me i dont actually know if ? inside syscalls transforms to an error code, or aborts. or if error codes have defined meanings for all syscalls or whether each syscall can define its own... in any event it would be good to define a way for the syscall to signal these error conditions to the implementation. inclined to say that if the sysvar is unavailable (if a sysvar is deprecated/removed, or specifically if we support EpochRewards) it should return an error code that the implementation can handle, but corrupt data or out of bounds read should abort

also im not sure what happens if the syscall attempts an out of bounds write. presumably it cant detect and prevent this because everything we do with this is unsafe transmute... at the least, the spec should say the caller must ensure their buffer is long enough

Reply:
> it occurs to me i dont actually know if ? inside syscalls transforms to an error code, or aborts.

It does indeed! It gets turned into a `u64` but then turned back into an `InstructionError` in the sysvar interface. So we have lots of flexibility there to define our own syscall-level errors, either with existing `InstructionError` variants or `InstructionError::Custom`.

> also im not sure what happens if the syscall attempts an out of bounds write.

The way I was thinking about this was that it would be an out of bounds _read_, and that's easy to detect because say you get `offset: 90` and your desired sysvar is only `60` bytes long, throw.